### DISCUSSION TITLE
Problem when using fatfs example in a out_of_tree_driver with the file ff.h

### DISCUSSION THREAD

**Discussion by Tobi15** [zephyrproject-rtos/zephyr#51399] [General] [answered]
## Describe the bug
ff.h file missing when trying to access from an out_of_tree_driver

## To Reproduce

I get t he sample "samples/application_development/out_of_tree_driver"
Then, I modified the driver with the "samples/subsys/fs/fatfs". I created an init() method in the C file and put the method signature in the H file.

In the main, I called this method init().

Run : "west build ..." 

## Expected behavior
The project can build properly

## Screenshots or console output

fatal error: ff.h: No such file or directory
   14 | #include <ff.h>
      |          ^~~~~~
compilation terminated.
ninja: build stopped: subcommand failed.

The error comes from the include in the driver H file.

## Source code 

[fat_fs_with_sd_driver.zip](https://github.com/zephyrproject-rtos/zephyr/files/9766345/fat_fs_with_sd_driver.zip)

**Comment by carlescufi** [comment]
@nordicjm @de-nordic any hints?

**Comment by nordicjm** [comment, marked as answer]
A driver added to zephyr externally is a driver and is added as an external module, it should have access to the base zephyr features, fat_fs is not a base zephyr feature, it's a module. Because they are both modules, you cannot reference it, and why would a driver expose a fat filesystem? The driver should expose a base device which is then mounted as a filesystem - be it fat or littlefs or any other filesystem, which can then be mounted from application code.

The reason for there being a split of modules that can be used and those that can't, from an external module, is because some modules configuration is in the zephyr tree, and some is outside of it. fatfs is outside of it, segger is inside it (which is why you can enable RTT output and it will work from a module)

**Comment by de-nordic** [comment]
The <ff.h> header is not part of zephyr interface unless you enable FAT FS support.
The FAT FS module adds its include paths with CMakeLists.txt line:
https://github.com/zephyrproject-rtos/fatfs/blob/a30531af3a95a9a3ea7d771ea8a578ebfed45514/CMakeLists.txt#L5
and ELMFAT is added to FS interface with line
https://github.com/zephyrproject-rtos/zephyr/blob/3cef8e68df3ede8dc9b41e76953ba74f87b2da7a/subsys/fs/CMakeLists.txt#L19
under condition that `CONFIG_FAT_FILESYSTEM_ELM=y`
If your driver links zephyr_interface then you should enable the FAT FS driver module, to get access to ff.h, but this will also enable default FAT FS driver, so probably cause collision with what you are doing.


Anyway: I do not understand why SD card driver requires FAT file system access. SD card driver should provide disk interface that can then be used to mount FAT FS of that interface.

**Comment by de-nordic** [comment]
I allowed myself to remove "bug" label, because this is not a bug.

**Comment by Tobi15** [comment]
Thank you for your answers! I didn't understand it before. I changed the way I solve my problem by using a library to do my work. 
Best regard!
